src/Stratis.Bitcoin.Features.SmartContracts/Rules/AllowedScriptTypeRule.cs

```diff
@@ -50,31 +50,15 @@ public static void CheckTransaction(Network network, Transaction transaction)
             }
         }
 
-        private static void CheckOutput(TxOut output)
+        private static void CheckOutput(Network network, TxOut output)
         {
             if (output.ScriptPubKey.IsSmartContractExec())
                 return;
 
             if (output.ScriptPubKey.IsSmartContractInternalCall())
                 return;
 
-            if (PayToPubkeyHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
-                return;
-
-            // For cross-chain transfers
-            if (PayToScriptHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
-                return;
-
-            // For cross-chain transfers
-            if (PayToMultiSigTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
-                return;
-
-            // For cross-chain transfers
-            if (PayToFederationTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
-                return;
-
-            // For cross-chain transfers
-            if (TxNullDataTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
+            if (network.StandardScriptsRegistry.IsStandardScriptPubKey(network, output.ScriptPubKey))
```

Comment:
I would revert this part, although the intention of the change is good. The purpose of these checks is that on the smart contract networks we require an address to be extractable from transactions (multisig transactions being an exception) so that the account model can be correctly layered on top of the UTXO model.

It is also not desirable to conflate consensus with standardness too much. As a rule of thumb the standardness rules are more restrictive than consensus. Therefore the standard scripts registry could diverge from what is allowed in this consensus rule (albeit a shared consensus and mempool rule).